{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Done! Congratulations on your new bot. You will find it at t.me/ArbolitoV2_bot. You can now add a description, about section and profile picture for your bot, see /help for a list of commands. By the way, when you've finished creating your cool bot, ping our Bot Support if you want a better username for it. Just make sure the bot is fully operational before you do this.\n",
    "\n",
    "Use this token to access the HTTP API:\n",
    "7933607614:AAE68SZk1Ia_1NP8l6-y9PsT8NhuN_Wzvdg\n",
    "Keep your token secure and store it safely, it can be used by anyone to control your bot.\n",
    "\n",
    "For a description of the Bot API, see this page: https://core.telegram.org/bots/api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip uninstall python-telegram-bot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Importamos Librerias\n",
    "from telegram.ext import Application, CommandHandler, Updater\n",
    "import logging  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Importamos el Token\n",
    "bot_token = \"7933607614:AAE68SZk1Ia_1NP8l6-y9PsT8NhuN_Wzvdg\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bot started. Press Ctrl+C to stop.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-24 16:06:41,017 - httpx - INFO - HTTP Request: POST https://api.telegram.org/bot7933607614:AAE68SZk1Ia_1NP8l6-y9PsT8NhuN_Wzvdg/getMe \"HTTP/1.1 200 OK\"\n",
      "2025-04-24 16:06:41,243 - httpx - INFO - HTTP Request: POST https://api.telegram.org/bot7933607614:AAE68SZk1Ia_1NP8l6-y9PsT8NhuN_Wzvdg/deleteWebhook \"HTTP/1.1 200 OK\"\n",
      "2025-04-24 16:06:41,246 - telegram.ext.Application - INFO - Application started\n",
      "2025-04-24 16:06:41,950 - httpx - INFO - HTTP Request: POST https://api.telegram.org/bot7933607614:AAE68SZk1Ia_1NP8l6-y9PsT8NhuN_Wzvdg/getUpdates \"HTTP/1.1 200 OK\"\n",
      "2025-04-24 16:06:42,304 - httpx - INFO - HTTP Request: POST https://api.telegram.org/bot7933607614:AAE68SZk1Ia_1NP8l6-y9PsT8NhuN_Wzvdg/sendMessage \"HTTP/1.1 200 OK\"\n",
      "2025-04-24 16:06:52,191 - httpx - INFO - HTTP Request: POST https://api.telegram.org/bot7933607614:AAE68SZk1Ia_1NP8l6-y9PsT8NhuN_Wzvdg/getUpdates \"HTTP/1.1 200 OK\"\n",
      "2025-04-24 16:06:52,424 - httpx - INFO - HTTP Request: POST https://api.telegram.org/bot7933607614:AAE68SZk1Ia_1NP8l6-y9PsT8NhuN_Wzvdg/getUpdates \"HTTP/1.1 200 OK\"\n",
      "2025-04-24 16:06:52,426 - telegram.ext.Application - INFO - Application is stopping. This might take a moment.\n",
      "2025-04-24 16:06:52,427 - telegram.ext.Application - INFO - Application.stop() complete\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Cannot close a running event loop",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[5], line 123\u001b[0m\n\u001b[0;32m    121\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    122\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBot started. Press Ctrl+C to stop.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m--> 123\u001b[0m     loop\u001b[38;5;241m.\u001b[39mrun_until_complete(\u001b[43mapplication\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_polling\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[0;32m    124\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[0;32m    125\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBot stopped.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\site-packages\\telegram\\ext\\_application.py:868\u001b[0m, in \u001b[0;36mApplication.run_polling\u001b[1;34m(self, poll_interval, timeout, bootstrap_retries, read_timeout, write_timeout, connect_timeout, pool_timeout, allowed_updates, drop_pending_updates, close_loop, stop_signals)\u001b[0m\n\u001b[0;32m    865\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21merror_callback\u001b[39m(exc: TelegramError) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    866\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcreate_task(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprocess_error(error\u001b[38;5;241m=\u001b[39mexc, update\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m--> 868\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m__run\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    869\u001b[0m \u001b[43m    \u001b[49m\u001b[43mupdater_coroutine\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mupdater\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstart_polling\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    870\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpoll_interval\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpoll_interval\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    871\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    872\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbootstrap_retries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbootstrap_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    873\u001b[0m \u001b[43m        \u001b[49m\u001b[43mread_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mread_timeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    874\u001b[0m \u001b[43m        \u001b[49m\u001b[43mwrite_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mwrite_timeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    875\u001b[0m \u001b[43m        \u001b[49m\u001b[43mconnect_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconnect_timeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    876\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpool_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpool_timeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    877\u001b[0m \u001b[43m        \u001b[49m\u001b[43mallowed_updates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mallowed_updates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    878\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdrop_pending_updates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdrop_pending_updates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    879\u001b[0m \u001b[43m        \u001b[49m\u001b[43merror_callback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merror_callback\u001b[49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# if there is an error in fetching updates\u001b[39;49;00m\n\u001b[0;32m    880\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mclose_loop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mclose_loop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    882\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstop_signals\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop_signals\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    883\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\site-packages\\telegram\\ext\\_application.py:1099\u001b[0m, in \u001b[0;36mApplication.__run\u001b[1;34m(self, updater_coroutine, stop_signals, close_loop)\u001b[0m\n\u001b[0;32m   1097\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m   1098\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m close_loop:\n\u001b[1;32m-> 1099\u001b[0m         \u001b[43mloop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclose\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\asyncio\\selector_events.py:101\u001b[0m, in \u001b[0;36mBaseSelectorEventLoop.close\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     99\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mclose\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m    100\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mis_running():\n\u001b[1;32m--> 101\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot close a running event loop\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    102\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mis_closed():\n\u001b[0;32m    103\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m\n",
      "\u001b[1;31mRuntimeError\u001b[0m: Cannot close a running event loop"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "from telegram.ext import Application, CommandHandler, MessageHandler, filters, ConversationHandler\n",
    "import pandas as pd\n",
    "import io\n",
    "import asyncio\n",
    "import nest_asyncio\n",
    "import os\n",
    "import importlib.util\n",
    "\n",
    "# Apply nest_asyncio to allow nested event loops\n",
    "nest_asyncio.apply()\n",
    "\n",
    "# Configure logging\n",
    "logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)\n",
    "\n",
    "# Define conversation states\n",
    "CHOOSING_AGENT = 0\n",
    "\n",
    "# Get the directory of the current script\n",
    "try:\n",
    "    BASE_DIR = os.path.dirname(os.path.abspath(__file__))\n",
    "except NameError:\n",
    "    current_dir = os.getcwd()\n",
    "    \n",
    "    BASE_DIR = os.path.dirname(current_dir)  # Use current working directory if __file__ is not defined\n",
    "\n",
    "# Define your command functions\n",
    "async def start(update, context):\n",
    "    await update.message.reply_text('Bienvenido/a! Por favor, elige un agente para iniciar el proceso (ej., \"BNA\"):')\n",
    "    return CHOOSING_AGENT\n",
    "\n",
    "async def process_agent(update, context):\n",
    "    agent = update.message.text.upper()\n",
    "    script_path = os.path.join(BASE_DIR, 'BNA', 'obtener_cotizacion.py')\n",
    "\n",
    "    logging.info(f\"Attempting to run script for agent {agent}\")\n",
    "    logging.info(f\"Looking for script at path: {script_path}\")\n",
    "\n",
    "    if not os.path.exists(script_path):\n",
    "        await update.message.reply_text(f\"Lo siento, no se encontró ningún agente llamado {agent} en {script_path}.\")\n",
    "        return ConversationHandler.END\n",
    "\n",
    "    # Send an intermediate message\n",
    "    wait_message = await update.message.reply_text(f\"Por favor espere, estoy procesando la solicitud para {agent}. Esto puede tomar unos segundos...\")\n",
    "\n",
    "    try:\n",
    "        # Import and run the script\n",
    "        spec = importlib.util.spec_from_file_location(\"module.name\", script_path)\n",
    "        module = importlib.util.module_from_spec(spec)\n",
    "        spec.loader.exec_module(module)\n",
    "\n",
    "        # Assuming the script has a get_exchange_rate function\n",
    "        result = module.get_exchange_rate()\n",
    "\n",
    "        if result:\n",
    "            # Read the CSV file\n",
    "            csv_path = os.path.join( 'exchange_rates.csv')\n",
    "            if not os.path.exists(csv_path):\n",
    "                await update.message.reply_text(f\"Error: CSV file not found at {csv_path}\")\n",
    "                return ConversationHandler.END\n",
    "\n",
    "            df = pd.read_csv(csv_path)\n",
    "            \n",
    "            # Convert DataFrame to CSV string\n",
    "            csv_buffer = io.StringIO()\n",
    "            df.to_csv(csv_buffer, index=False)\n",
    "            csv_buffer.seek(0)\n",
    "\n",
    "            # Delete the wait message\n",
    "            await wait_message.delete()\n",
    "\n",
    "            # Send the CSV file\n",
    "            await context.bot.send_document(\n",
    "                chat_id=update.effective_chat.id,\n",
    "                document=io.BytesIO(csv_buffer.getvalue().encode()),\n",
    "                filename='exchange_rates.csv',\n",
    "                caption=f'Aquí están los \"exchange rates\" para {agent}.'\n",
    "            )\n",
    "        else:\n",
    "            # Delete the wait message\n",
    "            await wait_message.delete()\n",
    "            await update.message.reply_text(f\"El script para {agent} se ejecutó, pero no devolvió ningún resultado.\")\n",
    "\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error running script for {agent}: {str(e)}\", exc_info=True)\n",
    "        # Delete the wait message\n",
    "        await wait_message.delete()\n",
    "        await update.message.reply_text(f\"Ocurrió un error al procesar la solicitud para {agent}: {str(e)}\")\n",
    "\n",
    "    return ConversationHandler.END\n",
    "\n",
    "async def cancel(update, context):\n",
    "    await update.message.reply_text('Operation cancelled.')\n",
    "    return ConversationHandler.END\n",
    "\n",
    "async def main():\n",
    "    # Create the application and pass your bot's token\n",
    "    application = Application.builder().token(bot_token).build()\n",
    "\n",
    "    # Add conversation handler\n",
    "    conv_handler = ConversationHandler(\n",
    "        entry_points=[CommandHandler('start', start)],\n",
    "        states={\n",
    "            CHOOSING_AGENT: [MessageHandler(filters.TEXT & ~filters.COMMAND, process_agent)]\n",
    "        },\n",
    "        fallbacks=[CommandHandler('cancel', cancel)]\n",
    "    )\n",
    "\n",
    "    application.add_handler(conv_handler)\n",
    "\n",
    "    return application\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    # Get the current event loop\n",
    "    loop = asyncio.get_event_loop()\n",
    "\n",
    "    # Create and configure the application\n",
    "    application = loop.run_until_complete(main())\n",
    "\n",
    "    # Start the bot\n",
    "    try:\n",
    "        print(\"Bot started. Press Ctrl+C to stop.\")\n",
    "        loop.run_until_complete(application.run_polling())\n",
    "    except KeyboardInterrupt:\n",
    "        print(\"Bot stopped.\")\n",
    "    finally:\n",
    "        loop.run_until_complete(application.shutdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Cannot close a running event loop",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\site-packages\\telegram\\ext\\_application.py:1069\u001b[0m, in \u001b[0;36mApplication.__run\u001b[1;34m(self, updater_coroutine, stop_signals, close_loop)\u001b[0m\n\u001b[0;32m   1068\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 1069\u001b[0m     \u001b[43mloop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_until_complete\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minitialize\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1070\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpost_init:\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\asyncio\\base_events.py:663\u001b[0m, in \u001b[0;36mBaseEventLoop.run_until_complete\u001b[1;34m(self, future)\u001b[0m\n\u001b[0;32m    662\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_closed()\n\u001b[1;32m--> 663\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_check_running\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    665\u001b[0m new_task \u001b[38;5;241m=\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m futures\u001b[38;5;241m.\u001b[39misfuture(future)\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\asyncio\\base_events.py:622\u001b[0m, in \u001b[0;36mBaseEventLoop._check_running\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    621\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mis_running():\n\u001b[1;32m--> 622\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mThis event loop is already running\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m    623\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m events\u001b[38;5;241m.\u001b[39m_get_running_loop() \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[1;31mRuntimeError\u001b[0m: This event loop is already running",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\site-packages\\telegram\\ext\\_application.py:1094\u001b[0m, in \u001b[0;36mApplication.__run\u001b[1;34m(self, updater_coroutine, stop_signals, close_loop)\u001b[0m\n\u001b[0;32m   1093\u001b[0m         loop\u001b[38;5;241m.\u001b[39mrun_until_complete(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpost_stop(\u001b[38;5;28mself\u001b[39m))\n\u001b[1;32m-> 1094\u001b[0m \u001b[43mloop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_until_complete\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshutdown\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1095\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpost_shutdown:\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\asyncio\\base_events.py:663\u001b[0m, in \u001b[0;36mBaseEventLoop.run_until_complete\u001b[1;34m(self, future)\u001b[0m\n\u001b[0;32m    662\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_closed()\n\u001b[1;32m--> 663\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_check_running\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    665\u001b[0m new_task \u001b[38;5;241m=\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m futures\u001b[38;5;241m.\u001b[39misfuture(future)\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\asyncio\\base_events.py:622\u001b[0m, in \u001b[0;36mBaseEventLoop._check_running\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    621\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mis_running():\n\u001b[1;32m--> 622\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mThis event loop is already running\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m    623\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m events\u001b[38;5;241m.\u001b[39m_get_running_loop() \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[1;31mRuntimeError\u001b[0m: This event loop is already running",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[5], line 113\u001b[0m\n\u001b[0;32m    110\u001b[0m     application\u001b[38;5;241m.\u001b[39mrun_polling()\n\u001b[0;32m    112\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;18m__name__\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__main__\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m--> 113\u001b[0m     \u001b[43mmain\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[5], line 110\u001b[0m, in \u001b[0;36mmain\u001b[1;34m()\u001b[0m\n\u001b[0;32m    107\u001b[0m application\u001b[38;5;241m.\u001b[39madd_handler(conv_handler)\n\u001b[0;32m    109\u001b[0m \u001b[38;5;66;03m# Start the Bot\u001b[39;00m\n\u001b[1;32m--> 110\u001b[0m \u001b[43mapplication\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_polling\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\site-packages\\telegram\\ext\\_application.py:868\u001b[0m, in \u001b[0;36mApplication.run_polling\u001b[1;34m(self, poll_interval, timeout, bootstrap_retries, read_timeout, write_timeout, connect_timeout, pool_timeout, allowed_updates, drop_pending_updates, close_loop, stop_signals)\u001b[0m\n\u001b[0;32m    865\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21merror_callback\u001b[39m(exc: TelegramError) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    866\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcreate_task(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprocess_error(error\u001b[38;5;241m=\u001b[39mexc, update\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m--> 868\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m__run\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    869\u001b[0m \u001b[43m    \u001b[49m\u001b[43mupdater_coroutine\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mupdater\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstart_polling\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    870\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpoll_interval\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpoll_interval\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    871\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    872\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbootstrap_retries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbootstrap_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    873\u001b[0m \u001b[43m        \u001b[49m\u001b[43mread_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mread_timeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    874\u001b[0m \u001b[43m        \u001b[49m\u001b[43mwrite_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mwrite_timeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    875\u001b[0m \u001b[43m        \u001b[49m\u001b[43mconnect_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconnect_timeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    876\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpool_timeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpool_timeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    877\u001b[0m \u001b[43m        \u001b[49m\u001b[43mallowed_updates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mallowed_updates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    878\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdrop_pending_updates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdrop_pending_updates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    879\u001b[0m \u001b[43m        \u001b[49m\u001b[43merror_callback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merror_callback\u001b[49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# if there is an error in fetching updates\u001b[39;49;00m\n\u001b[0;32m    880\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mclose_loop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mclose_loop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    882\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstop_signals\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop_signals\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    883\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\site-packages\\telegram\\ext\\_application.py:1099\u001b[0m, in \u001b[0;36mApplication.__run\u001b[1;34m(self, updater_coroutine, stop_signals, close_loop)\u001b[0m\n\u001b[0;32m   1097\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m   1098\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m close_loop:\n\u001b[1;32m-> 1099\u001b[0m         \u001b[43mloop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclose\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\s1093678\\AppData\\Local\\miniconda3\\Lib\\asyncio\\selector_events.py:101\u001b[0m, in \u001b[0;36mBaseSelectorEventLoop.close\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     99\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mclose\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m    100\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mis_running():\n\u001b[1;32m--> 101\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot close a running event loop\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    102\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mis_closed():\n\u001b[0;32m    103\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m\n",
      "\u001b[1;31mRuntimeError\u001b[0m: Cannot close a running event loop"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import logging\n",
    "import importlib.util\n",
    "import pandas as pd\n",
    "import io\n",
    "from telegram import Update\n",
    "from telegram.ext import Application, CommandHandler, MessageHandler, filters, ConversationHandler\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "# Define the base directory where your script is located\n",
    "BASE_DIR = r\"C:\\Users\\s1093678\\OneDrive - Syngenta\\Documents\\Crop Protection\\Cursos\\Master en Data Science\\20 - Web Mining\\arbolito\\BNA\"\n",
    "\n",
    "# Define conversation states\n",
    "CHOOSING = 0\n",
    "\n",
    "async def start(update: Update, context):\n",
    "    await update.message.reply_text(\n",
    "        \"Bienvenido al bot de cotizaciones de bancos. \"\n",
    "        \"Por favor, elija un banco (BNA, PROVINCIA, CIUDAD) o escriba 'TODOS' para obtener todas las cotizaciones.\"\n",
    "    )\n",
    "    return CHOOSING\n",
    "\n",
    "async def process_bank(update: Update, context):\n",
    "    bank = update.message.text.upper()\n",
    "    script_path = os.path.join(BASE_DIR, 'obtener_cotizacion_v2.py')\n",
    "\n",
    "    logging.info(f\"Attempting to run script for bank {bank}\")\n",
    "    logging.info(f\"Looking for script at path: {script_path}\")\n",
    "\n",
    "    if not os.path.exists(script_path):\n",
    "        await update.message.reply_text(f\"Lo siento, no se encontró el script en {script_path}.\")\n",
    "        return ConversationHandler.END\n",
    "\n",
    "    # Send an intermediate message\n",
    "    wait_message = await update.message.reply_text(f\"Por favor espere, estoy procesando la solicitud para {bank}. Esto puede tomar unos segundos...\")\n",
    "\n",
    "    try:\n",
    "        # Import and run the script\n",
    "        spec = importlib.util.spec_from_file_location(\"module.name\", script_path)\n",
    "        module = importlib.util.module_from_spec(spec)\n",
    "        spec.loader.exec_module(module)\n",
    "\n",
    "        # Call the appropriate function based on the bank\n",
    "        if bank == 'BNA':\n",
    "            result = module.get_exchange_rate_BNA()\n",
    "        elif bank == 'PROVINCIA':\n",
    "            result = module.get_exchange_rate_banco_provincia()\n",
    "        elif bank == 'CIUDAD':\n",
    "            result = module.get_exchange_rate_bancociudad()\n",
    "        elif bank == 'TODOS':\n",
    "            result = (module.get_exchange_rate_BNA() and \n",
    "                      module.get_exchange_rate_banco_provincia() and \n",
    "                      module.get_exchange_rate_bancociudad())\n",
    "        else:\n",
    "            await update.message.reply_text(f\"Banco no reconocido: {bank}\")\n",
    "            return ConversationHandler.END\n",
    "\n",
    "        # Check if the CSV file was created\n",
    "        csv_path = os.path.join(BASE_DIR, 'exchange_rates.csv')\n",
    "        if not os.path.exists(csv_path):\n",
    "            logging.error(f\"CSV file not created at {csv_path}\")\n",
    "            await update.message.reply_text(f\"Error: No se pudo crear el archivo CSV en {csv_path}\")\n",
    "            return ConversationHandler.END\n",
    "\n",
    "        # Read the CSV file\n",
    "        df = pd.read_csv(csv_path)\n",
    "        \n",
    "        # Convert DataFrame to CSV string\n",
    "        csv_buffer = io.StringIO()\n",
    "        df.to_csv(csv_buffer, index=False)\n",
    "        csv_buffer.seek(0)\n",
    "\n",
    "        # Delete the wait message\n",
    "        await context.bot.delete_message(chat_id=update.effective_chat.id, message_id=wait_message.message_id)\n",
    "\n",
    "        # Send the CSV file\n",
    "        await context.bot.send_document(\n",
    "            chat_id=update.effective_chat.id,\n",
    "            document=io.BytesIO(csv_buffer.getvalue().encode()),\n",
    "            filename='exchange_rates.csv',\n",
    "            caption=f'Aquí están los \"exchange rates\" para {bank}.'\n",
    "        )\n",
    "\n",
    "    except Exception as e:\n",
    "        logging.error(f\"Error running script for {bank}: {str(e)}\", exc_info=True)\n",
    "        # Delete the wait message\n",
    "        await context.bot.delete_message(chat_id=update.effective_chat.id, message_id=wait_message.message_id)\n",
    "        await update.message.reply_text(f\"Ocurrió un error al procesar la solicitud para {bank}: {str(e)}\")\n",
    "\n",
    "    return ConversationHandler.END\n",
    "\n",
    "def main():\n",
    "    # Replace 'YOUR_BOT_TOKEN' with your actual bot token\n",
    "    application = Application.builder().token('7933607614:AAE68SZk1Ia_1NP8l6-y9PsT8NhuN_Wzvdg').build()\n",
    "\n",
    "    conv_handler = ConversationHandler(\n",
    "        entry_points=[CommandHandler('start', start)],\n",
    "        states={\n",
    "            CHOOSING: [MessageHandler(filters.TEXT & ~filters.COMMAND, process_bank)],\n",
    "        },\n",
    "        fallbacks=[],\n",
    "    )\n",
    "\n",
    "    application.add_handler(conv_handler)\n",
    "\n",
    "    # Start the Bot\n",
    "    application.run_polling()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
